From cd530d3d49918a9105b6797e72dba04819af302a Mon Sep 17 00:00:00 2001
From: Sebastian Ullrich <sebasti@nullri.ch>
Date: Wed, 15 Sep 2021 13:16:12 +0200
Subject: [PATCH] link: windows-gnu: look for Unix-style library names

---
 src/link/Coff.zig | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/src/link/Coff.zig b/src/link/Coff.zig
index 4f5df73f8d..19553ee455 100644
--- a/src/link/Coff.zig
+++ b/src/link/Coff.zig
@@ -1297,11 +1297,16 @@ fn linkWithLLD(self: *Coff, comp: *Compilation) !void {
                 continue;
             }
             if (target.abi.isGnu()) {
-                const fallback_name = try allocPrint(arena, "lib{s}.dll.a", .{key});
+                const fallback_name = try allocPrint(arena, "lib{s}.a", .{key});
                 if (try self.findLib(arena, fallback_name)) |full_path| {
                     argv.appendAssumeCapacity(full_path);
                     continue;
                 }
+                const fallback_implib_name = try allocPrint(arena, "lib{s}.dll.a", .{key});
+                if (try self.findLib(arena, fallback_implib_name)) |full_path| {
+                    argv.appendAssumeCapacity(full_path);
+                    continue;
+                }
             }
             log.err("DLL import library for -l{s} not found", .{key});
             return error.DllImportLibraryNotFound;

